Mathlib/Analysis/Calculus/FirstDerivativeTest.lean

```diff
@@ -0,0 +1,176 @@
+    (g₀ : a < b) (g₁ : b < c)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b)) (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ : ∀ x ∈ Set.Ioo a b, deriv f x ≤ 0)
+    (h₁ : ∀ x ∈ Set.Ioo b c, 0 ≤ deriv f x) : IsLocalMin f b := by
+    have Q := @first_derivative_test_Ioo_max (-f) a b c g₀ g₁
+      (by simp_all) (DifferentiableOn.neg hd₀) (DifferentiableOn.neg hd₁)
+      (fun _ => deriv_neg_nonneg hd₀ h₀) (fun _ => deriv_neg_nonpos hd₁ h₁)
+    unfold IsLocalMax IsMaxFilter at Q
+    simp only [Pi.neg_apply, neg_le_neg_iff] at Q
+    exact Q
+
+
+ /-- The First-Derivative Test from calculus, maxima version,
+ expressed in terms of left and right filters. -/
+lemma first_derivative_test_max₀ {f : ℝ → ℝ} {b : ℝ} (h : ContinuousAt f b)
+    (hd₀ : ∀ᶠ x in 𝓝[<] b, DifferentiableAt ℝ f x) (hd₁ : ∀ᶠ x in 𝓝[>] b, DifferentiableAt ℝ f x)
+    (h₀  : ∀ᶠ x in 𝓝[<] b, 0 ≤ deriv f x) (h₁  : ∀ᶠ x in 𝓝[>] b, deriv f x ≤ 0) :
+    IsLocalMax f b := by
+  unfold Filter.Eventually at h₀ h₁ hd₀ hd₁
+  rw [mem_nhdsWithin_Iio_iff_exists_Ioo_subset] at h₀ hd₀
+  rw [mem_nhdsWithin_Ioi_iff_exists_Ioo_subset] at h₁ hd₁
+  obtain ⟨u₀,hu₀⟩ := hd₀; obtain ⟨u₁,hu₁⟩ := hd₁
+  obtain ⟨v₀,hv₀⟩ := h₀; obtain ⟨v₁,hv₁⟩ := h₁
+  apply isLocalMax_of_mono_anti
+  · show max u₀ v₀ < b; exact max_lt (by simp_all) (by simp_all)
+  · show b < min u₁ v₁; exact lt_min (by simp_all) (by simp_all)
+  · exact monotoneOn_of_deriv_nonneg
+      (convex_Ioc _ _)
+      (fun x _ => ContinuousAt.continuousWithinAt ((em (x = b)).elim (fun H => H ▸ h)
+        (fun H => DifferentiableAt.continuousAt (hu₀.2 (by contrapose H;simp_all;linarith)))))
+      (fun x _ => DifferentiableAt.differentiableWithinAt (hu₀.2 (by simp_all)))
+      (fun x _ => by apply hv₀.2;simp_all)
+  · exact antitoneOn_of_deriv_nonpos
+      (convex_Ico _ _)
+      (fun x _ => ContinuousAt.continuousWithinAt ((em (x = b)).elim (fun H => H ▸ h)
+        (fun H => DifferentiableAt.continuousAt (hu₁.2 (by contrapose H;simp_all;linarith)))))
+      (fun x _ => DifferentiableAt.differentiableWithinAt (hu₁.2 (by simp_all)))
+      (fun x _ => by apply hv₁.2;simp_all)
+
+/-- If a set `P` contains left and right neighborhoods of a point `x`
+then `P` contains a punctured neighborhood. -/
+lemma nhdsWithin_punctured_of_Iio_Ioi (P : Set ℝ)
+    (hl : P ∈ nhdsWithin 0 (Set.Iio 0)) (hr : P ∈ nhdsWithin 0 (Set.Ioi 0)) :
+    P ∈ nhdsWithin 0 {0}ᶜ := by
+  rw [mem_nhdsWithin] at *
+  obtain ⟨u,hu⟩ := hl
+  obtain ⟨v,hv⟩ := hr
+  exact ⟨u ∩ v, IsOpen.inter (by tauto) (by tauto), by tauto,
+    fun x hx => by
+      simp_all only [Set.mem_inter_iff]
+      exact  ((lt_or_gt_of_ne hx.2)).elim (fun _ => hu.2.2 (by tauto)) (fun _ => hv.2.2 (by tauto))⟩
+
+/-- If a set `P` contains a punctured neighborhood of `x`
+then `P` contains a left neighborhoods of `x`. -/
+lemma nhdsWithin_Iio_of_punctured {b:ℝ} {P : Set ℝ} (h : P ∈ nhdsWithin b {b}ᶜ) :
+    P ∈ nhdsWithin b (Set.Iio b) := by
```

Comment:
This is `  nhds_left'_le_nhds_ne b h`, no need for a new lemma.

Reply:
Thanks, fixed.